tcpchecker.go

```diff
@@ -0,0 +1,137 @@
+package checkup
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"time"
+)
+
+// TCPChecker implements a Checker for TCP endpoints.
+type TCPChecker struct {
+	// Name is the name of the endpoint.
+	Name string `json:"endpoint_name"`
+
+	// URL is the URL of the endpoint.
+	URL string `json:"endpoint_url"`
+
+	// TLSEnabled controls whether to enable TLS or not.
+	// If set, TLS is enabled.
+	TLSEnabled bool `json:"tls,omitempty"`
+
+	// TLSSkipVerify controls whether to skip server TLS
+	// certificat validation or not.
+	TLSSkipVerify bool `json:"tls_skip_verify,omitempty"`
+
+	// TLSCAFile is the Certificate Authority used
+	// to validate the server TLS certificate.
+	TLSCAFile string `json:"tls_ca_file,omitempty"`
+
+	// Timeout is the maximum time to wait for a
+	// TCP connection to be established.
+	Timeout time.Duration `json:"timeout,omitempty"`
+
+	// ThresholdRTT is the maximum round trip time to
+	// allow for a healthy endpoint. If non-zero and a
+	// request takes longer than ThresholdRTT, the
+	// endpoint will be considered unhealthy. Note that
+	// this duration includes any in-between network
+	// latency.
+	ThresholdRTT time.Duration `json:"threshold_rtt,omitempty"`
+
+	// Attempts is how many requests the client will
+	// make to the endpoint in a single check.
+	Attempts int `json:"attempts,omitempty"`
+}
+
+// Check performs checks using c according to its configuration.
+// An error is only returned if there is a configuration error.
+func (c TCPChecker) Check() (Result, error) {
+	if c.Attempts < 1 {
+		c.Attempts = 1
+	}
+
+	result := Result{Title: c.Name, Endpoint: c.URL, Timestamp: Timestamp()}
+	result.Times = c.doChecks()
+
+	return c.conclude(result), nil
+}
+
+// doChecks executes and returns each attempt.
+func (c TCPChecker) doChecks() Attempts {
+	var err error
+	var conn net.Conn
+
```

Comment:
Instead of changing the struct above for a timeout, consider just doing this:

```go
timeout := c.Timeout
if timeout == 0 {
    timeout = 5 * time.Second
}
```

Or something. Then use `timeout` instead of `c.Timeout` directly.